Out of tree (driver) configuration support

thomasstenersen opened the pull request:
# Summary

This PR mainly defines and adds support for out-of-source drivers. Specifically in the context of using the BLE controller in https://github.com/NordicPlayground/nrfxlib and https://github.com/NordicPlayground/fw-nrfconnect-nrf, but the approach should be generic for other similar use cases where an out-of-source subsystem needs specific control over certain drivers/peripherals.

In addition the PR has some minor fixes for making Zephyr play nicer with "outside" configuration.

https://github.com/zephyrproject-rtos/zephyr/pull/13631/commits/8f1e6cb177e2f8fc49b0830589439d92c4ecf260  should provide some explanation and reasoning for the main changes in this PR.

## FAQ

Answering the question you might have when seeing this PR.

> Do we really need this? Isn't there another way?

Yes we need this, unfortunately. This is for the case were an out-of-tree component for some reason _must_ use it's own driver. There is no way in Kconfig that I and others have found to effectively de-select a Kconfig choice/config without severely polluting the Zephyr configuration files.

> Are we going to have such a kconfig for each driver subsystem?

I would add it at a per use-case/need basis. But this is a generic pattern, yes.

> Are we talking about 2 conflicting drivers and we would select the one out of tree instead of the one in the tree? 

Yes. And the fact that they are not only conflicting but another subsystem _must_ use the out-of-tree one. We also want this switch to happen automatically, so if I select a different BLE controller in the menuconfig, the correct driver dependencies are resolved without the need for other changes by the user.

> In short, this looks like a hack to me.

I would like it for there to be a nicer way of doing this. But as far as I can tell, this is the most pragmatic, least intrusive and nicest way of handling out-of-tree dependencies like this. IMO the only way of actually solving the root problem is to resolve https://github.com/zephyrproject-rtos/zephyr/issues/8181.

Commit:
kconfig: Use depend on instead of select to avoid kconfig loop

Remove use of select to "force" enabling other configs in subsys/fs
and subsys/net/l2. The forcing will cause infinite kconfig recursion.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Commit:
kconfig: bluetooth: Let BT_HCI_VS be configurable from controller

This moves the vendor-specific HCI command/event configuration
definitions out of bluetooth/common into bluetooth. This allows
the controller itself to indicate its support for vendor-specific
commands/events.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Commit:
drivers: nrf5: Don't force specific drivers from soc/arm/nordic_nrf

Redefining the config will not let another (out-of-source) driver be
chosen instead of the default. The driver is practically forced by the
soc settings. This commit moves default settings from soc/arm/nordic_nrf
into the drivers themselves.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Commit:
drivers: clock_control: nrf: add choice variable

Add choice variables for CLOCK_CONTROL_NRF_SOURCE and
CLOCK_CONTROL_NRF_ACCURACY such that the choices may be augmented
out-of-tree.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Commit:
drivers: add support for forcing out-of-tree drivers

An external project extending the Zephyr RTOS and its drivers may have
subsystems that must use its own specific driver(s) when active. One
example is the nRF5x NVMC that must be scheduled in between radio
operations. A subsystem may also be dependent on its own drivers for
security, real-time and/or because of hardware constrains.

In order to not introduce non-Zephyr specific code into the Zephyr tree,
an option is added to disable the in-tree drivers in Zephyr. Because
Kconfig does not support a good way of de-selecting other symbols, a
variable on the form `<DRIVER>_MUST_USE_OUT_OF_TREE_DRIVER` is added as
a dependency for each `<DRIVER>`. For example, the out-of-tree subsystem
will select `FLASH_MUST_USE_OUT_OF_TREE_DRIVER` to disable all in-tree
drivers. A solution for issue #8181 would open up for a more general
solution, however #8181 requires significant effort.

Support for out-of-tree drivers is added to clock_control, entropy and
flash.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Commit:
bluetooth: host: Specify controller in BT_HCI_TX_STACK_SIZE defaults

By specifying the controller directly when selecting the default
BT_HCI_TX_STACK_SIZE, an external controller may rely on the final
default value when none of the in-tree controllers are used.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

SebastianBoe commented on drivers/entropy/CMakeLists.txt:
```diff
@@ -1,10 +1,8 @@
+zephyr_sources_ifdef(CONFIG_ENTROPY_SAM_RNG            entropy_sam.c)
+zephyr_sources_ifdef(CONFIG_ENTROPY_STM32_RNG          entropy_stm32.c)
+zephyr_sources_ifdef(CONFIG_FAKE_ENTROPY_NATIVE_POSIX  fake_entropy_native_posix.c)
+zephyr_sources_ifdef(CONFIG_USERSPACE                  entropy_handlers.c)
```
I undertand that this scheme requires drivers to not use zephyr_library, but (albeit not for entropy) there are valid use-cases for using zephyr_library in drivers.

So we need to do something else.

I believe that if we merge #13791, then out-of-tree drivers can do

target_sources(drivers__entropy PRIVATE my/out/of/tree/driver.c)

Then we don't need to re-organize the drivers, and we don't need to avoid using zephyr_library_* in drivers.

thomasstenersen replied:
Sounds like that could work and is the sane way of doing this. I'll check out this locally and test if it works.

thomasstenersen replied:
Looks like the approach in #13791 works fine for this :+1: Anything stopping that from being merged?

SebastianBoe replied:
I'm certain it will be merged. It's blocked by review, review is going a bit slow these days due to embedded world.

thomasstenersen replied:
Ok, I removed the changes to the particular CMakeLists.txt file and we'll use the patch in #13791 in the out-of-source tree.

SebastianBoe commented on drivers/clock_control/Kconfig:
```diff
@@ -9,6 +9,14 @@
 # Clock controller drivers
 #
+
+config CLOCK_CONTROL_HAS_SUBSYS_DRIVER
```
I think we want a different name for this option.

It's especially the 'SUBSYS' part that I'm concnerned about. I understand that in this particular case, an out-of-tree subsystem needed to force a particular driver.

But it might be in the future that an out-of-tree driver, or out-of-tree app might want to force their driver. So the 'subsys' term doesn't work. Perhaps 

CLOCK_CONTROL_HAS_OOT_DRIVER
CLOCK_CONTROL_HAS_FORCED_DRIVER
CLOCK_CONTROL_HAS_OUTSIDE_DRIVER

I'd rather not use the term 'external' as it is not as precise as out-of-tree.

thomasstenersen replied:
I agree, the name is not the best. I don't mind long names here, so something `CLOCK_CONTROL_HAS_OUT_OF_TREE_DRIVER` is fine by me. Other ideas/suggestions are welcome as well.

b0661 replied:
I assume "CLOCK_CONTROL_HAS_SUBSYS_DRIVER" and the "if  CLOCK_CONTROL_HAS_SUBSYS_DRIVER"  guard was introduced to allow the OOT driver and the Zephyr driver to use the same Kconfig symbols without interfering each other.

In my view the usage of the same Kconfig symbols should be controlled by the driver. The Zephyr driver or the OOT driver should enable the Kconfig symbols by a non user selectable Kconfig option (e.g. CLOCK_CONTROL_NRF_HAS_DRIVER).

A driver that wants to activate the options just has to select CLOCK_CONTROL_xxx_HAS_DRVER.

```
menuconfig CLOCK_CONTROL_MY_NRF
	bool "My NRF Clock controller support"
	depends on SOC_COMPATIBLE_NRF
        select CLOCK_CONTROL_NRF_HAS_DRIVER
	help
	  Enable my specific support for the Nordic Semiconductor nRFxx series SoC clock driver.
```

```
menuconfig CLOCK_CONTROL_NRF
	bool "NRF Clock controller support"
	depends on SOC_COMPATIBLE_NRF
        select CLOCK_CONTROL_NRF_HAS_DRIVER
	help
	  Enable support for the Nordic Semiconductor nRFxx series SoC clock
	  driver.

if CLOCK_CONTROL_NRF_HAS_DRIVER

config CLOCK_CONTROL_NRF_IRQ_PRIORITY
	int "Power Clock Interrupt Priority"
	range 0 7
	default 1
	help
	  The interrupt priority for Power Clock interrupt.

  ...
```


That would also allow to keep the Kconfig option for the driver mostly in one place.

thomasstenersen replied:
> I assume "CLOCK_CONTROL_HAS_SUBSYS_DRIVER" and the "if CLOCK_CONTROL_HAS_SUBSYS_DRIVER" guard was introduced to allow the OOT driver and the Zephyr driver to use the same Kconfig symbols without interfering each other.

It should be `if !CLOCK_CONTROL_HAS_SUBSYS_DRIVER`.
This is not the reason. The driving use-case for this is for out-of-tree subsystems (or anything, really) to have a way of de-selecting in-tree drivers. This could be an external subsystem that is Zephyr agnostic and needs all flash, RNG, etc. operations to be done through its own APIs. Keeping the drivers tied to a particular subsystem could, for example, be to meet real-time and/or security requirements or because of hardware constraints.

Furthermore, we don't want to pollute the Zephyr config files, drivers, etc. with out-of-tree specific configuration and code.

Let's say that I have a subsystem `FOO` that when selected needs to use it's own driver `FLASH_DRIVER_FOO`. The problem arises when the user selects `CONFIG_FOO=y`, `FOO` has no way of saying that, e.g., `FLASH_DRIVER_NRF` is incompatible with it. 

I hope this cleared things a bit up, let me know if there is something I missed or didn't explain well enough.

SebastianBoe replied:
Let's go for CLOCK_CONTROL_HAS_OUT_OF_TREE_DRIVER

SebastianBoe replied:
... or maybe 

CLOCK_CONTROL_MUST_USE_OUT_OF_TREE_DRIVER

HAS implies that it is available to use, but in reality, the option indicates that the driver must be used.

thomasstenersen replied:
I agree on the semantics of HAS vs MUST, but I'd argue that naming consistency is more valuable. `<DRIVER>_HAS_OUT_OF_TREE_DRIVER` it is,

SebastianBoe replied:
But what about when in the future we HAVE an out-of-tree driver, but don't want to force it to be enabled?

thomasstenersen replied:
If you have a driver but it's not enforced, you'd add it to a choice selection. Then you can select driver at will and you shouldn't use this mechanism. I don't think that is an argument here.

By consistent naming I'm talking about using `HAS` to resolve `depends on` dependencies in a sane way. I would like to keep different names to a minimum, i.e., remove usage of `SUPPORTS` and the like. The I don't have to ask myself every time: "was it `<FOO>_HAS_<BAR>` or `<FOO>_SUPPORTS_<BAR>`, ..." 

That being said, _I do acknowledge that this case is different_ though, is it different enough to warrant its own convention?

SebastianBoe replied:
Yes.

thomasstenersen replied:
Changed

SebastianBoe commented on drivers/flash/Kconfig:
```diff
@@ -45,6 +52,7 @@ config FLASH_PAGE_LAYOUT
 	help
 	  Enables API for retrieving the layout of flash memory pages.
 
+if !FLASH_HAS_SUBSYS_DRIVER
```
I'd like to avoid wrapping all the in-tree Kconfig sources with this. But I suspect we have no choice ... I'll get back to you about our options.

thomasstenersen replied:
It's not the best, but this is the cleanest option I've found.

SebastianBoe replied:
Yeah, I can't find anything cleaner either.

This will have to do until we have better Kconfig tooling.

thomasstenersen commented on drivers/clock_control/Kconfig:
```diff
@@ -9,6 +9,14 @@
+
+config CLOCK_CONTROL_HAS_SUBSYS_DRIVER
+  bool
+  default n
```
Sure
- [ ] Remove `default n`

thomasstenersen replied:
Removed

thomasstenersen commented on drivers/clock_control/Kconfig:
```diff
@@ -9,6 +9,14 @@
+  bool
+  default n
+  help
+		This option can be enabled by an external library if it
```
Good idea, I'll do a `grep`

thomasstenersen replied:
Renamed instances of "external library" to "out-of-tree library".

thomasstenersen commented on drivers/flash/Kconfig:
```diff
@@ -41,6 +41,7 @@ config FLASH_SHELL
 config FLASH_PAGE_LAYOUT
 	bool "API for retrieving the layout of pages"
 	depends on FLASH_HAS_PAGE_LAYOUT
+	default y
```
Looks like my `git` wizardry failed me. I'll fix up the commit message

thomasstenersen replied:
Fixed.

thomasstenersen commented on drivers/flash/Kconfig:
```diff
@@ -41,6 +41,7 @@ config FLASH_SHELL
 config FLASH_PAGE_LAYOUT
 	bool "API for retrieving the layout of pages"
 	depends on FLASH_HAS_PAGE_LAYOUT
+	default y
```
I agree :broken_heart:  Alas, with the current limitations, I'm not sure if it is better to default to true here or force all the examples to explicitly enable all their required features. I fear the latter option will require a lot of work :cry:

thomasstenersen commented on soc/arm/nordic_nrf/Kconfig.defconfig:
```diff
@@ -12,20 +12,6 @@ source "soc/arm/nordic_nrf/*/Kconfig.defconfig.series"
-	default y
-
-endif # ENTROPY_GENERATOR
-
```
Any force-selection of drivers is not good IMO, so if we find more instances of that, it should be changed. Usage of `select` on variables not related to indicating support for something (`*_HAS_*`) seems to be an instant Kconfig spaghetti creator.

SebastianBoe replied:
Yes, I think you are right about that ...

PavelVPV commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -8,39 +8,6 @@
 
 if BT_HCI
 
-config BT_HCI_VS
```
Initially, BT_HCI_VS was defined in `subsys/bluetooth/common/Kconfig`. But the LL options are applied later in `subsys/bluetooth/controller/Kconfig`. See `subsys/bluetooth/Kconfig`:
```
source "subsys/bluetooth/common/Kconfig"
source "subsys/bluetooth/host/Kconfig"
source "subsys/bluetooth/controller/Kconfig"
source "subsys/bluetooth/shell/Kconfig"
```
In case when LL doesn't support Zephyr's VS commands and events, there is no way to say it, because LL configuration is defined after this configuration.  More over, BT_HCI_VS is enabled by default.

Therefore, in order to control this configuration from LL, this configuration is moved down after the LL configuration. So now LL configuration can say: select BT_HAS_HCI_VS, which would mean that it supports BT_HCI_VS

SebastianBoe replied:
> In case when LL doesn't support Zephyr's VS commands and events, there is no way to say it, because LL configuration is defined after this configuration.

I don't understand how the order can affect things.

My claim is that we can make do with introducing the *HAS* option, we don't need to move things around.

```
It's perfectly valid to do:

config BT_HAS_HCI_VS
	bool

config BT_HCI_VS
	bool "Zephyr HCI Vendor-Specific Commands"
	depends on BT_HAS_HCI_VS
	default y

[...]

select BT_HAS_HCI_VS

```

PavelVPV replied:
Checked again. True. It isn't needed anymore sinse `BT_HAS_HCI_VS` is added. This code block should be moved back to common/Kconfig together with `BT_HAS_HCI_VS`

thomasstenersen replied:
Fixed.

thomasstenersen commented on subsys/bluetooth/Kconfig:
```diff
@@ -144,6 +144,47 @@ source "subsys/bluetooth/common/Kconfig"
+	bool
+	help
+	  This option can be enabled by an external library if it
+	  supports the Zephyr HCI Vendor-Specific Commands and Events.
```
It technically doesn't have to, it's optional to indicate support. How about:

> This option is set by the Bluetooth controller to indicate support for the Zephyr HCI Vendor-Specific Commands and Event.

thomasstenersen replied:
Done.

thomasstenersen commented on drivers/clock_control/Kconfig:
```diff
@@ -9,6 +9,14 @@
+  bool
+  default n
+  help
+		This option can be enabled by an external library if it
```
Fixed.

Vudentz commented on subsys/net/l2/Kconfig:
```diff
@@ -16,11 +16,11 @@ config NET_L2_DUMMY
+	depends on BT_CENTRAL
+	depends on BT_SMP
+	depends on BT_L2CAP_DYNAMIC_CHANNEL
 	select NET_6LO
```
Don't know exactly why it would go into an infine loop, expect perhaps if there is a circular dependecy somewhere, but it really sounds bad that it apparently on affect BT_ options not NET_ options so application are forced to mark the 5 extra options.

thomasstenersen replied:
It is a circular dependency, starting at BT_CTRL => BT => NET_BUF .... => BT_CTRL again ([kconfig-loop.txt](https://github.com/zephyrproject-rtos/zephyr/files/2918449/kconfig-loop.txt)). And the problem goes deeper than here, really. There are many more places selecting symbols in a way that's prone to these problems. IMO since the implementation of Kconfig (that resolves the symbols) can't resolve dependencies automagically (see issue #8181) we have to force users to be explicit. It is better that Kconfig fails with a message [1]:

> waning: FOO (defined in ..) was assigned the value 'y' but got the value 'n'

than having a huge circular dependency to fix because of symbols selecting each other uncritically. 

[1] This error message could be improved though, it could say
> FOO was assigned 'y' but got the value 'n' because one or more of its dependencies were not met.
> FOO depends on BAR (=n) && BAZ (=y)

Vudentz replied:
> It is a circular dependency, starting at BT_CTRL => BT => NET_BUF .... => BT_CTRL again ([kconfig-loop.txt](https://github.com/zephyrproject-rtos/zephyr/files/2918449/kconfig-loop.txt)).

I guess the dependecy is BT_CTRL -> ... -> BT_LL_CHOISE -> BT_CTRL, I don't see how NET_BUF would cause this since it has no dependencies something else is causing this circular depency. In fact I think this is related to BT_CTRL as NET_L2_BT don't have a depency to it (btw it seems you guys are trying to enable openthread so I wonder why NET_L2_BT or you want both 15.4 an BT L2 drivers?)

thomasstenersen replied:
If you have a another/better way of dissolving this circular dependency I am happy to add it. I don't think there is only one here. I simply removed using `select` by net/l2. I am not trying to enable openthread, I just want to resolve the circular dependency.

Do you have a suggestion for what I should do instead?

Vudentz replied:
Can you paste the .conf file in use in this case, anyway there got to be a way to fix circular dependencies somehow, perhaps switch to imply or you tried that already?

thomasstenersen replied:
`imply` does not change anything, unfortunately.

[.config](https://github.com/zephyrproject-rtos/zephyr/files/2919499/dot_config.txt)

SebastianBoe commented:
@thoh-ot : Hi, does Oticon also need an out-of-tree subsystem to force certain drivers to be enabled?

If so, would Oticon be able to review this PR and see if it meets Oticon's requirements?

aescolar commented:
> @thoh-ot : Hi, does Oticon also need an out-of-tree subsystem to force certain out-of-tree drivers to be enabled?

Yes, we certainly do. @kapedev should be looking into this.

aescolar commented on soc/arm/nordic_nrf/nrf51/Kconfig.series:
```diff
@@ -12,7 +12,6 @@ config  SOC_SERIES_NRF51X
 	select SOC_FAMILY_NRF
 	select NRF_RTC_TIMER
 	select CLOCK_CONTROL
-	select CLOCK_CONTROL_NRF
```
For consistency it would be then nice to also update the nrf52_bsim Kconfig.board

thomasstenersen replied:
Updated

aescolar commented:
@SebastianBoe 
Overall: this PR is ok for us in the sense that it does not harm our needs, and may help if we need to avoid some in-tree drivers kconfig files being included.
But it does not yet address the main issue.
That main issue being how do we get out of tree folders to be part of what cmake and Kconfig go thru (similar to the BOARD/SOR/ARCH_ROOT options), but for any arbitrary set of component.
There is some thoughts here about the possibility of having a way to define an "overlay" folder structure, that could contain all the needed extra components following the same folder structure as in the zephyr root, but only containing the folders that need replacing/extending (so what cmake and kconfig would traverse would be something like an union of both the zephyr root and that overlay root)
As said, hopefully @kapedev will soon engage with this. It is very nice to see you guys have similar needs, hopefully we will be having a nice solution that covers both of our needs.

SebastianBoe commented:
> That main issue being how do we get out of tree folders to be part of what cmake and Kconfig go thru (similar to the BOARD/SOR/ARCH_ROOT options)

Note that this recently became possible through west. Out-of-tree CMake and Kconfig code can be sourced as if it were an in-tree subsys. An example of this in action can be seen below.

This patch to west.yaml:

https://github.com/zephyrproject-rtos/zephyr/pull/13672/files#diff-4ae7a6d2c692e706a33f7a71d3eb25d2R38

causes the files (zephyr/CMakeLists.txt and zephyr/Kconfig) in another repo to be add_subdirectory'd / sourced:

https://github.com/JuulLabs-OSS/mcuboot/pull/430/files#diff-83aeef700a5a192d68e562d9ccfe3725R1

But this is not quite as powerful as the overlay you describe, as you can't inject out-of-tree CMake/Kconfig code into a specific in-tree CMake directory scope, or an in-tree Kconfig source-hierarchy.

An overlay mechanism would be interesting, but I'm not sure how one could implement it in a sane way.

thomasstenersen commented:
Fixed merge conflicts and updated nrf52_bsim Kconfig.board as @aescolar pointed out. Is there anything left for me to do here or can we merge this @aescolar @Vudentz?

aescolar commented:
@thomasstenersen : I'm ok with it. But I do not own/maintain most of it. So I let the approvals for those who do.

carlescufi commented:
I am fine with this except for the notation:

`*_MUST_USE_OUT_OF_TREE_DRIVER` is simply too long and verbose.
I propose either:

* `*_API_ONLY`
* `*_EXT_DRIVER`
* `*_OOT_DRIVER`

Thoughts?

ulfalizer commented on drivers/clock_control/Kconfig:
```diff
@@ -9,6 +9,13 @@
+  help
+	  This option can be enabled by an out-of-tree library if it
+	  needs to take control over the clock controller.
+
```
Weird formatting. Check how other symbols are formatted, with tabs vs. spaces, etc.

thomasstenersen replied:
It is inconsistent across many, many files unfortunately. As far as I've seen, most places use tab + 2x space:
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/clock_control/Kconfig.nrf#L11
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/counter/Kconfig.qmsi#L13
https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/Kconfig#L20

ulfalizer replied:
Thinking about before the `bool` and the `help`. There's two spaces there now instead of a tab.

ulfalizer commented on drivers/clock_control/Kconfig:
```diff
@@ -23,6 +30,8 @@ module = CLOCK_CONTROL
 source "subsys/logging/Kconfig.template.log_config"
 
+
+if !CLOCK_CONTROL_MUST_USE_OUT_OF_TREE_DRIVER
```
Bit unbalanced to have no blank line after the `if` but a blank line before the `endif`.

I like blank lines for both. Less scrunched up.

thomasstenersen replied:
I'm usually a stickler for consistency. I can fix this and let the config breathe.

ulfalizer commented on drivers/entropy/Kconfig:
```diff
@@ -6,19 +6,27 @@
+  bool
+  help
+	  This option can be enabled by an out-of-tree library if it
+	  needs to take control over the entropy driver.
```
Weird formatting.

thomasstenersen replied:
What is weird? Tab + spaces?

ulfalizer replied:
Two spaces before `bool` and `help` instead of a tab. Makes it render wonky on GitHub too.

ulfalizer commented on drivers/entropy/Kconfig:
```diff
@@ -6,19 +6,27 @@
 
-if ENTROPY_GENERATOR
 
+if ENTROPY_GENERATOR
```
Bit hard to read. Looks like the `if ENTROPY_GENERATOR` goes just with the `source`s, due to the scrunching up.

thomasstenersen replied:
You want to add a few more blank lines then? Around the two if's and the endif?

ulfalizer replied:
Thinking it could be e.g.

```
if X

if Y
blah
blah
endif # Y

endif # X
```

instead of

```
if X
if Y
blah
blah
endif # Y

endif # X
```

Second one kinda makes it look like the `if X` only attaches to the first block.

ulfalizer commented on drivers/flash/Kconfig:
```diff
@@ -20,6 +20,12 @@ config FLASH_HAS_PAGE_LAYOUT
+  bool
+  help
+	  This option can be enabled by an out-of-tree library if it
+	  needs to take control over the flash memory peripheral.
```
Weird formatting.

ulfalizer commented on drivers/flash/Kconfig:
```diff
@@ -65,4 +72,5 @@ source "drivers/flash/Kconfig.sam"
 
 source "drivers/flash/Kconfig.w25qxxdv"
 
+endif # !FLASH_MUST_USE_OUT_OF_TREE_DRIVER
```
Inconsistent blank lines.

ulfalizer commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -48,17 +48,19 @@ config BT_RX_BUF_LEN
 
 config BT_HCI_TX_STACK_SIZE
+	# NOTE: By _not_ having a prompt here, Kconfig will fall back on
+	# the default value if none of the other values are set.
```
Could say something like "Since this symbol isn't user-configurable (has no prompt), it will always get its value from the defaults."

There's a lot of those comments strewn randomly in the Kconfig files though. I wonder how much they help in practice. Maybe it would be better to remove all of them and add more documentation on how prompts work somewhere else.

The [Kconfig - Tips and Best Practices](https://docs.zephyrproject.org/latest/tools/kconfig/index.html) page has some handy tips btw.

thomasstenersen replied:
I've read all the manuals I could come across :+1: I generally agree that duplication of information like this is a bad thing. In this case, the note is there to warn anyone that making this a prompt will break the automagic selection of stack size depending on controller selected etc. That's the _real_ problem that we're trying to solve, but I don't know any method that can actually safe-guard against this without putting a lot of burden on the user :/

ulfalizer replied:
Maybe this would work, if you think there's a risk someone might add a prompt later: "This value is derived from other symbols and should not be user-configurable. Do not give it a prompt."

ulfalizer reviewed (changes requested):
Just some random style nits (on vacation :). Could take a closer look later.

I think we should be really careful so that we don't accommodate out-of-tree Kconfig configuration too much in the main repo though. It adds enough complexity that there has to be a significant advantage.

Keeping the main repo simple and understandable is probably more important on the whole than making life easy for people doing out-of-tree stuff. Linux has the right balance there I think.

ulfalizer commented:
If someone doesn't mind being inconsistent and wants to indent items within `if`s and `choice`s, etc., I'll approve it by the way.

It makes Kconfig much easier to read and less error-prone, especially when you have nested `if`s. Shame it wasn't what people did in the beginning. Seen it in some projects at least.

mped-oticon commented on soc/arm/nordic_nrf/nrf52/Kconfig.series:
```diff
@@ -13,7 +13,6 @@ config  SOC_SERIES_NRF52X
 	select SOC_FAMILY_NRF
 	select NRF_RTC_TIMER
 	select CLOCK_CONTROL
-	select CLOCK_CONTROL_NRF
```
91 and 51 appear to be addressed now.

thomasstenersen commented:
@ulfalizer Thanks for taking the time to review the changes. I believed I've addressed all the comments with the latest push.

thomasstenersen commented:
@ulfalizer Thanks for taking the time to review the changes. I believed I've addressed all the comments with the latest push.

ulfalizer commented on drivers/clock_control/Kconfig:
```diff
@@ -9,6 +9,13 @@
 # Clock controller drivers
 #
+
+config CLOCK_CONTROL_MUST_USE_OUT_OF_TREE_DRIVER
```
Maybe this could be called `CLOCK_CONTROL_OUT_OF_TREE_DRIVER` or the like, to shorten it a bit. It might be clear anyway that there must be an out-of-tree driver if it's on.

thomasstenersen replied:
(@carlescufi) We had a naming discussion here: https://github.com/zephyrproject-rtos/zephyr/pull/13631#discussion_r260318620

The reason for the `_MUST_USE_` (originally `_HAS_`)  was to keep consistency with other select-able Kconfig symbols. If we keep select-able symbols on the form `<noun>_<verb>_<noun>` (`FOO_HAS_BAR`), it is much easier to keep your sanity when working with the Kconfig files.

The name is verbose, but I'd rather sacrifice verbosity for clarity.

SebastianBoe replied:
The problem with OUT_OF_TREE_DRIVER is that we need some option/way to distinguish between the scenario where there exists an out-of-tree driver that may be selected, and the scenario where there exists an out-of-tree driver that must be selected.

And OUT_OF_TREE_DRIVER does not reflect this.

carlescufi replied:
If you really want to enforce the `<noun>_<verb>_<noun>` (which I don't think we could call a convention right now except for the `_HAS_` case) then I would go for something like `*_SELECT_OOT` or `REQUIRE_OOT` something like that. `MUST_USE` sounds convoluted and verbose to me.

SebastianBoe replied:
The original proposal of _HAS_ didn't make sense, but now I'm unsure if we actually considered using an option without HAS. As Ulf says, if the OOT driver is enabled, that does imply that the other drivers can't be. The symbol for an enabled out-of-tree driver would look like:

CLOCK_CONTROL_OUT_OF_TREE_DRIVER

ulfalizer reviewed (commented):
I'm a noob on the Bluetooth stuff, but getting rid of some `select`s seems nice.

Maybe things could be refactored further so that you don't have to enable lots of different things to be able to turn on `NET_L2_BT` (worth checking if all combinations of values for those `depends on`'ed symbols make sense), but that's unrelated.

nashif commented on drivers/clock_control/Kconfig:
```diff
@@ -9,6 +9,13 @@
 # Clock controller drivers
 #
+
+config CLOCK_CONTROL_MUST_USE_OUT_OF_TREE_DRIVER
```
hmmm, why do we need to have Kconfigs for this? Are we going to have such a kconfig for each driver subsystem? Are we talking about 2 conflicting drivers and we would select the one out of tree instead  of the one in the tree? In short, this looks like a hack to me. Need some documentation or a design proposal explaining how this can be done in a generic way without polluting Kconfig with such options.

SebastianBoe replied:
> Need some documentation or a design proposal explaining how this can be done in a generic way without polluting Kconfig with such options.

Please prove me wrong @ulfalizer, it is unavoidable to pollute the kconfig with these options without resolving #8181, which is not going to happen any time soon.

For your other questions, please see https://github.com/zephyrproject-rtos/zephyr/pull/13631/commits/8f1e6cb177e2f8fc49b0830589439d92c4ecf260

carlescufi replied:
The reasoning is described in this commit message: https://github.com/zephyrproject-rtos/zephyr/pull/13631/commits/8f1e6cb177e2f8fc49b0830589439d92c4ecf260

thomasstenersen replied:
Yes we need this, unfortunately. This is for the case were an out-of-tree component for some reason _must_ use it's own driver. There is no way in Kconfig that I and others have found to effectively de-select a Kconfig choice/config without severely polluting the Zephyr configuration files.

> Are we going to have such a kconfig for each driver subsystem?

I would add it at a per use-case/need basis. But this is a generic pattern, yes.

> Are we talking about 2 conflicting drivers and we would select the one out of tree instead of the one in the tree? 

Yes. And the fact that they are not only conflicting but another subsystem _must_ use the out-of-tree one. We also want this switch to happen automatically, so if I select a different BLE controller in the menuconfig, the correct driver dependencies are resolved without the need for other changes by the user.

> In short, this looks like a hack to me.

I would like it for there to be a nicer way of doing this. But as far as I can tell, this is the most pragmatic, least intrusive and nicest way of handling out-of-tree dependencies like this. IMO the only way of actually solving the root problem is to resolve https://github.com/zephyrproject-rtos/zephyr/issues/8181.

> Need some documentation or a design proposal [...]

What would such a proposal look like?  Could I add more to https://github.com/zephyrproject-rtos/zephyr/pull/13631/commits/8f1e6cb177e2f8fc49b0830589439d92c4ecf260 to explain the why and how better?

carlescufi commented:
@aescolar 
> That main issue being how do we get out of tree folders to be part of what cmake and Kconfig go thru (similar to the BOARD/SOR/ARCH_ROOT options), but for any arbitrary set of component.

As @SebastianBoe already mentioned, this is already possible *with or without* west. You can either use west to list the modules you need or you can provide the build system with a list of them directly via the `ZEPHYR_MODULES` variable: https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/zephyr_module.cmake#L11

thomasstenersen commented:
Closing PR to rework changes to be more Nordic specific.
I'll create _two_ new PRs. One for the various minor fixups and one for the out-of-tree driver support itself.
